File: packages/core/src/modules/hassault/decals.ts
/**
 * Bullet marks: where the shots went, still there when you look back.
 *
 * The world used to keep no memory of being shot at. A tracer for 75ms, a blob
 * at the endpoint for 300ms, and then nothing — so a firefight left a room
 * looking exactly as it had before it, and a player had no way to see their own
 * spray on a wall. Since the spray is now a *learnable pattern*
 * (`weapons.py`'s `spray`), the marks are not decoration: they are how you read
 * the pattern you are supposed to be learning.
 *
 * ## The normal comes off the wire
 *
 * A decal has to lie flat on the surface it is on, which means knowing which
 * face was hit. That is **not** derived here: `weapons.raycast_world_face`
 * already knows it at the instant its walk returns, and puts one small integer
 * per pellet into the `shot` fx. Working it out here instead would be a copy of
 * the world ray whose only job is to agree with the server about the exact point
 * the server chose — and a disagreement of half a cube puts the mark inside the
 * wall, where it is invisible and reports nothing. See `FACE_NORMALS`.
 *
 * ## A ring buffer, not a list
 *
 * `DECAL_MAX` meshes are allocated once and reused in order. Nothing is created
 * or destroyed per shot, and the cap holds *by construction* rather than by a
 * `while (length > MAX) shift()` that has to run after every push. An automatic
 * weapon puts thirty marks on a wall in two seconds; that is a rate at which
 * per-shot allocation is a real cost.
 *
 * Takes `three` as a parameter rather than importing it, like `effects.ts` and
 * `viewmodel.ts`, so the lazy-load stays in one place.
 */
import type * as THREE from 'three';

import { FACE_NONE } from './trace';

/** How many marks the world remembers. The 129th retires the 1st. */
export const DECAL_MAX = 128;
/** How long one mark lasts before it starts fading, in seconds. */
export const DECAL_LIFE = 22;
/** How long the fade at the end of that life takes. */
export const DECAL_FADE = 4;
/**
 * How far off the surface a mark sits, in cube units.
 *
 * A decal coplanar with the wall it is on z-fights, which reads as flicker
 * rather than as a bug. Lifted along the face normal by well under a millimetre
 * of world scale — `polygonOffset` does most of the work and this covers the
 * grazing angles where it does not.
 */
export const DECAL_LIFT = 0.012;
/** The mark's width, in cube units. A cube is roughly 36cm, so this is ~7cm. */
export const DECAL_SIZE = 0.2;

/**
 * The six faces as unit vectors, in **cube** coordinates.
 *
 * Indexed by the face on the wire, and mirroring `weapons.FACE_NORMALS` on the
 * server and `FACE_NORMALS` in `decals.rs`. Pinned by `browser_parity.rs`.
 */
export const FACE_NORMALS: readonly [number, number, number][] = [
  [1, 0, 0],
  [-1, 0, 0],
  [0, 1, 0],
  [0, -1, 0],
  [0, 0, 1],
  [0, 0, -1],
];

/** Tile edge in pixels. A bullet hole seen from across a room is a smudge. */
const SIZE = 32;

/** Deterministic hash in [0, 1), so a mark's cracks are the same every run. */
function hash(x: number, y: number): number {
  const n = Math.sin(x * 127.1 + y * 311.7) * 43758.5453123;
  return n - Math.floor(n);
}

/**
 * One impact tile: a dark crater, a lighter rim of displaced material, and a few
 * radial cracks.
 *
 * Alpha is the shape and the RGB is the colour, so one tile serves every
 * surface — a per-material tint is then one field on the material rather than a
 * second texture.
 */
export function drawImpactTile(size = SIZE): Uint8Array {
  const out = new Uint8Array(size * size * 4);
  const half = size / 2;
  for (let py = 0; py < size; py++) {
    for (let px = 0; px < size; px++) {
      const dx = (px + 0.5 - half) / half;
      const dy = (py + 0.5 - half) / half;
      const r = Math.hypot(dx, dy);
      const angle = Math.atan2(dy, dx);

      // The crater: solid to about a third of the radius, then falling away.
      let alpha = r < 0.34 ? 1 : Math.max(0, 1 - (r - 0.34) / 0.5);
      // Cracks: a handful of spokes reaching further out than the crater does,
      // roughened so the mark is not a target symbol.
      const spoke = Math.abs(Math.cos(angle * 2.5 + hash(px >> 3, py >> 3) * 2));
      alpha = Math.max(alpha, spoke > 0.93 && r < 0.92 ? 0.45 * (1 - r) : 0);
      // Hard zero at the rim, for the same reason the flash tile has one: a
      // decal with a visible edge reads as a sticker.
      if (r >= 1) alpha = 0;

      // The crater is nearly black and the rim just outside it is pale, which is
      // what makes a hole read as a hole rather than as a dark circle.
      const rim = r > 0.3 && r < 0.46 ? 1 : 0;
      const value = Math.round(rim ? 168 : 26);
      const i = (py * size + px) * 4;
      out[i] = value;
      out[i + 1] = value;
      out[i + 2] = value;
      out[i + 3] = Math.round(Math.max(0, Math.min(1, alpha)) * 255);
    }
  }
  return out;
}

/** Build the impact texture. A `DataTexture`, so this module runs headless. */
export function createImpactTexture(three: typeof THREE, size = SIZE): THREE.DataTexture {
  const texture = new three.DataTexture(drawImpactTile(size), size, size, three.RGBAFormat);
  texture.magFilter = three.LinearFilter;
  texture.minFilter = three.LinearMipmapLinearFilter;
  texture.generateMipmaps = true;
  texture.colorSpace = three.SRGBColorSpace;
  texture.needsUpdate = true;
  return texture;
}

/**
 * How opaque a mark is, `age` seconds after it was made.
 *
 * Flat for most of its life and then fading, rather than decaying from the
 * start: a mark that begins fading immediately is a mark that is never quite
 * legible, and legibility over a whole magazine is the entire point.
 */
export function decalOpacity(age: number): number {
  if (age <= DECAL_LIFE - DECAL_FADE) return 1;
  if (age >= DECAL_LIFE) return 0;
  return (DECAL_LIFE - age) / DECAL_FADE;
}

interface Mark {
  mesh: THREE.Mesh;
  material: THREE.MeshBasicMaterial;
  age: number;
  live: boolean;
}

export class DecalPool {
  private marks: Mark[] = [];
  private next = 0;
  private geometry: THREE.PlaneGeometry;
  private texture: THREE.DataTexture;
  /**
   * The six orientations, built once.
   *
   * **Fixed quaternions, never `lookAt`.** `lookAt`'s axis convention differs
   * between cameras and everything else, which is the kind of thing that is
   * wrong by exactly 180° — and a decal wrong by 180° is a decal facing into its
   * own wall, i.e. invisible. `effects.ts` documents the same trap for tracers.
   */
  private orientations: THREE.Quaternion[];

  constructor(
    // Not kept: the pool builds everything it needs here and then only moves
    // it, so holding the module would be a field nothing reads.
    three: typeof THREE,
    private readonly scene: THREE.Scene,
  ) {
    this.geometry = new three.PlaneGeometry(1, 1);
    this.texture = createImpactTexture(three);

    // A `PlaneGeometry` faces +z in its own space, so each orientation is the
    // rotation taking +z onto that face's normal — in **three**'s axes, where
    // cube z is up and cube y is three's z.
    const from = new three.Vector3(0, 0, 1);
    this.orientations = FACE_NORMALS.map((n) => {
      const to = new three.Vector3(n[0], n[2], n[1]);
      return new three.Quaternion().setFromUnitVectors(from, to);
    });

    this.marks = Array.from({ length: DECAL_MAX }, () => {
      const material = new three.MeshBasicMaterial({
        map: this.texture,
        transparent: true,
        opacity: 1,
        // Depth *tested* so a mark is hidden by anything in front of it, but
        // never *written*: marks overlap constantly and a written depth would
        // make the newer one z-fight the older.
        depthWrite: false,
        depthTest: true,
        polygonOffset: true,
        polygonOffsetFactor: -4,
        polygonOffsetUnits: -4,
      });
      const mesh = new three.Mesh(this.geometry, material);
      mesh.visible = false;
      // Below the view model's 2, so a mark can never draw over the gun.
      mesh.renderOrder = 1;
      scene.add(mesh);
      return { mesh, material, age: 0, live: false };
    });
  }

  /**
   * Record one impact, in **cube** coordinates — the axes the netcode speaks.
   *
   * A `face` of `FACE_NONE` is a body hit or a shot that ran out of range, and
   * is silently ignored: there is no surface to mark. Checked here rather than
   * at the call site so every caller has one fewer chance to index
   * `FACE_NORMALS` with `-1`.
   */
  mark(at: [number, number, number], face: number, tint = 0xffffff): void {
    if (face === FACE_NONE || face < 0 || face >= FACE_NORMALS.length) return;
    const entry = this.marks[this.next];
    this.next = (this.next + 1) % DECAL_MAX;

    const normal = FACE_NORMALS[face];
    // Cube (x, y, height) -> three (x, height, z), lifted off the surface along
    // its own normal so it does not z-fight the wall it is on.
    entry.mesh.position.set(
      at[0] + normal[0] * DECAL_LIFT,
      at[2] + normal[2] * DECAL_LIFT,
      at[1] + normal[1] * DECAL_LIFT,
    );
    entry.mesh.quaternion.copy(this.orientations[face]);
    // Spun about its own normal, so twenty marks on one wall are not twenty
    // copies of the same stamp.
    entry.mesh.rotateZ(hash(at[0], at[1]) * Math.PI * 2);
    entry.mesh.scale.setScalar(DECAL_SIZE);
    entry.material.color.setHex(tint);
    entry.material.opacity = 1;
    entry.mesh.visible = true;
    entry.age = 0;
    entry.live = true;
  }

  /** Age every live mark, fading and retiring as it goes. */
  update(dt: number): void {
    for (const entry of this.marks) {
      if (!entry.live) continue;
      entry.age += dt;
      const opacity = decalOpacity(entry.age);
      if (opacity <= 0) {
        entry.live = false;
        entry.mesh.visible = false;
        continue;
      }
      entry.material.opacity = opacity;
    }
  }

  /** How many marks are currently on the world. For tests and the debug HUD. */
  get size(): number {
    return this.marks.reduce((n, m) => n + (m.live ? 1 : 0), 0);
  }

  dispose(): void {
    for (const entry of this.marks) {
      this.scene.remove(entry.mesh);
      entry.material.dispose();
    }
    this.marks = [];
    this.geometry.dispose();
    this.texture.dispose();
  }
}

File: apps/native-fps/src/decals.rs
//! Bullet marks: where the shots went, still there when you look back.
//!
//! The browser's `decals.ts`, in this client's terms. Both hold a fixed ring of
//! marks, both take the surface from the wire rather than deriving it, and
//! `browser_parity.rs` pins the cap, the lifetime and the six face normals.
//!
//! ## The normal comes off the wire
//!
//! A mark has to lie flat on the surface it is on, which means knowing which
//! face was hit. `weapons.raycast_world_face` on the server already knows it at
//! the instant its walk returns, and puts one small integer per pellet into the
//! `shot` fx. Deriving it here would be a copy of the world ray whose only job
//! is to agree with the server about the exact point the server chose — and half
//! a cube of disagreement puts the mark inside the wall, invisible.
//!
//! ## Why it draws into the volume pass
//!
//! A decal is alpha-blended, must be occluded by anything in front of it, and
//! must not write depth (marks overlap constantly, and a written depth makes the
//! newer one z-fight the older). Those are exactly the three properties the
//! translucent volume pipeline is built for — the same one water and smoke use —
//! so this needs no pipeline, no bind group and no shader of its own.
//!
//! **It shares that pass's vertex budget**, which is the one thing to watch: a
//! wall covered in marks must not silently evict the smoke. `MAX_MARKS` is small
//! enough that it cannot (128 marks is 768 vertices against 65536), and the
//! pool is written first so that if the budget ever does bite it is the
//! newest-arriving effects that overflow rather than the world's memory of being
//! shot at.

use crate::trace::{FACE_NONE, FACE_NORMALS};

/// How many marks the world remembers. The 129th retires the 1st.
pub const MAX_MARKS: usize = 128;
/// How long one mark lasts before it starts fading, in seconds.
pub const DECAL_LIFE: f32 = 22.0;
/// How long the fade at the end of that life takes.
pub const DECAL_FADE: f32 = 4.0;
/// How far off the surface a mark sits, in cube units. A decal coplanar with its
/// wall z-fights, which reads as flicker rather than as a bug.
pub const DECAL_LIFT: f32 = 0.012;
/// The mark's width, in cube units. A cube is roughly 36cm, so this is ~7cm.
pub const DECAL_SIZE: f32 = 0.2;

/// The crater's colour. Near-black rather than black, so a mark on a dark
/// surface is still a mark and not a hole in the render.
const MARK_COLOR: [f32; 3] = [0.102, 0.102, 0.102];

/// One impact, in cube coordinates.
#[derive(Debug, Clone, Copy)]
struct Mark {
    at: [f32; 3],
    face: usize,
    age: f32,
    live: bool,
}

impl Default for Mark {
    fn default() -> Mark {
        Mark {
            at: [0.0; 3],
            face: 0,
            age: 0.0,
            live: false,
        }
    }
}

/// How opaque a mark is, `age` seconds after it was made.
///
/// Flat for most of its life and then fading, rather than decaying from the
/// start: a mark that begins fading immediately is never quite legible, and
/// legibility over a whole magazine is the entire point now that the spray is a
/// pattern to be learned.
pub fn decal_opacity(age: f32) -> f32 {
    if age <= DECAL_LIFE - DECAL_FADE {
        1.0
    } else if age >= DECAL_LIFE {
        0.0
    } else {
        (DECAL_LIFE - age) / DECAL_FADE
    }
}

/// A fixed ring of marks. Nothing is allocated per shot and the cap holds by
/// construction rather than by a trim that has to run after every push.
pub struct DecalPool {
    marks: [Mark; MAX_MARKS],
    next: usize,
}

impl Default for DecalPool {
    fn default() -> DecalPool {
        DecalPool {
            marks: [Mark::default(); MAX_MARKS],
            next: 0,
        }
    }
}

impl DecalPool {
    /// Record one impact.
    ///
    /// A `face` of [`FACE_NONE`] — a body hit, or a shot that ran out of range —
    /// is silently ignored: there is no surface to mark. Checked here rather
    /// than at the call site so every caller has one fewer chance to index
    /// `FACE_NORMALS` with `-1`.
    pub fn mark(&mut self, at: [f32; 3], face: i32) {
        if face == FACE_NONE || face < 0 || face as usize >= FACE_NORMALS.len() {
            return;
        }
        self.marks[self.next] = Mark {
            at,
            face: face as usize,
            age: 0.0,
            live: true,
        };
        self.next = (self.next + 1) % MAX_MARKS;
    }

    /// Every pellet of one shot, skipping the ones that found no surface.
    ///
    /// An empty `faces` marks nothing at all — a shooter whose backend predates
    /// the field leaves no marks, which is degraded rather than wrong. Guessing
    /// a face would put every one of them on the same arbitrary axis.
    pub fn shot(&mut self, ends: &[[f32; 3]], faces: &[i32]) {
        for (i, end) in ends.iter().enumerate() {
            self.mark(*end, faces.get(i).copied().unwrap_or(FACE_NONE));
        }
    }

    /// Age every live mark, retiring the ones that have faded out.
    pub fn update(&mut self, dt: f32) {
        for mark in self.marks.iter_mut() {
            if !mark.live {
                continue;
            }
            mark.age += dt;
            if mark.age >= DECAL_LIFE {
                mark.live = false;
            }
        }
    }

    /// Forget everything. A new map is a new world, and marks left over from the
    /// last one would hang in mid-air.
    pub fn clear(&mut self) {
        self.marks = [Mark::default(); MAX_MARKS];
        self.next = 0;
    }

    /// How many marks are currently on the world.
    pub fn len(&self) -> usize {
        self.marks.iter().filter(|m| m.live).count()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Two triangles per mark, lifted off their surface and squared to it.
    ///
    /// `push` takes the same `(position, color, alpha)` shape the volume pass
    /// wants; the caller supplies it so this module needs no renderer type and
    /// stays unit-testable headless, exactly as `effects.rs` does.
    ///
    /// **Positions come out in render coordinates**, like every other producer
    /// feeding this pass (`effects.rs`, `nades.rs`, `water.rs`). The wire, the
    /// face normals and the quad's own arithmetic are all cube space — `x`, `y`
    /// horizontal and `z` up — and the renderer is y-up, so the mapping happens
    /// once, here, at the push. Emitting cube coordinates instead does not fail:
    /// every mark simply lands with its height and its depth swapped, which
    /// buries it inside the geometry and reads as decals not working at all.
    pub fn quads(&self, mut push: impl FnMut([f32; 3], [f32; 3], f32)) {
        for mark in self.marks.iter() {
            if !mark.live {
                continue;
            }
            let alpha = decal_opacity(mark.age);
            if alpha <= 0.0 {
                continue;
            }
            let n = FACE_NORMALS[mark.face];
            let (u, v) = basis(n);
            let h = DECAL_SIZE * 0.5;
            let centre = [
                mark.at[0] + n[0] * DECAL_LIFT,
                mark.at[1] + n[1] * DECAL_LIFT,
                mark.at[2] + n[2] * DECAL_LIFT,
            ];
            let corner = |su: f32, sv: f32| {
                [
                    centre[0] + u[0] * su * h + v[0] * sv * h,
                    centre[1] + u[1] * su * h + v[1] * sv * h,
                    centre[2] + u[2] * su * h + v[2] * sv * h,
                ]
            };
            let a = corner(-1.0, -1.0);
            let b = corner(1.0, -1.0);
            let c = corner(1.0, 1.0);
            let d = corner(-1.0, 1.0);
            // Two triangles. Wound the same way for both, so a mark is visible
            // from the side its surface faces and from nowhere else — which is
            // what stops a floor mark showing through the ceiling below it.
            //
            // Cube (x, y, z-up) → render (x, height, y), the same mapping
            // `nades.rs` and `effects.rs` apply at their own push.
            for p in [a, b, c, a, c, d] {
                push([p[0], p[2], p[1]], MARK_COLOR, alpha);
            }
        }
    }
}

/// Two unit vectors spanning the plane a normal is perpendicular to.
///
/// The axis-aligned case only, because a face normal is always an axis: picking
/// the two remaining axes is exact and needs no cross product, no normalisation
/// and no degenerate case to guard.
fn basis(n: [f32; 3]) -> ([f32; 3], [f32; 3]) {
    if n[0] != 0.0 {
        ([0.0, 1.0, 0.0], [0.0, 0.0, 1.0])
    } else if n[1] != 0.0 {
        ([1.0, 0.0, 0.0], [0.0, 0.0, 1.0])
    } else {
        ([1.0, 0.0, 0.0], [0.0, 1.0, 0.0])
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::trace::{FACE_NX, FACE_PZ};

    #[test]
    fn a_body_hit_leaves_no_mark() {
        // The wall behind the body was never reached, and a mark on it would be
        // a lie about where the shot went.
        let mut pool = DecalPool::default();
        pool.mark([1.0, 2.0, 3.0], FACE_NONE);
        assert!(pool.is_empty());
    }

    #[test]
    fn an_out_of_range_face_is_refused_rather_than_indexed() {
        // `FACE_NONE` is negative so this cannot happen by accident, but a
        // malformed or newer server is not something to panic on.
        let mut pool = DecalPool::default();
        pool.mark([0.0; 3], 99);
        pool.mark([0.0; 3], -7);
        assert!(pool.is_empty());
    }

    #[test]
    fn a_shot_with_no_faces_marks_nothing() {
        // An older backend says nothing about surfaces. Marking nothing is
        // degraded; guessing a face would put every mark on the same axis.
        let mut pool = DecalPool::default();
        pool.shot(&[[1.0, 1.0, 1.0], [2.0, 2.0, 2.0]], &[]);
        assert!(pool.is_empty());
    }

    #[test]
    fn each_pellet_gets_its_own_face() {
        let mut pool = DecalPool::default();
        pool.shot(
            &[[1.0, 0.0, 0.0], [2.0, 0.0, 0.0], [3.0, 0.0, 0.0]],
            &[FACE_NX, FACE_NONE, FACE_PZ],
        );
        assert_eq!(pool.len(), 2);
    }

    #[test]
    fn the_ring_retires_the_oldest_rather_than_growing() {
        let mut pool = DecalPool::default();
        for i in 0..MAX_MARKS * 3 {
            pool.mark([i as f32, 0.0, 0.0], FACE_NX);
        }
        assert_eq!(pool.len(), MAX_MARKS);
    }

    #[test]
    fn a_mark_stays_legible_and_then_goes() {
        assert_eq!(decal_opacity(0.0), 1.0);
        assert_eq!(decal_opacity(DECAL_LIFE - DECAL_FADE), 1.0);
        assert!((decal_opacity(DECAL_LIFE - DECAL_FADE * 0.5) - 0.5).abs() < 1e-5);
        assert_eq!(decal_opacity(DECAL_LIFE), 0.0);

        let mut pool = DecalPool::default();
        pool.mark([0.0; 3], FACE_NX);
        pool.update(DECAL_LIFE + 0.1);
        assert!(pool.is_empty());
    }

    #[test]
    fn a_mark_comes_out_in_render_coordinates() {
        // The one thing every other producer feeding the volume pass does and
        // this one did not: cube (x, y, z-up) → render (x, height, y). Without
        // it a mark's height and depth are swapped, so it is drawn inside the
        // map instead of on it — and every other assertion here still passed,
        // because they were all written in cube space too.
        let mut pool = DecalPool::default();
        pool.mark([10.0, 20.0, 30.0], FACE_PZ);
        let mut out = Vec::new();
        pool.quads(|p, _, _| out.push(p));
        assert_eq!(out.len(), 6);
        for p in out {
            // x is untouched, the mark's own height (30 + the lift) is render y,
            // and the cube's y is render z.
            assert!((p[0] - 10.0).abs() <= DECAL_SIZE, "x moved: {p:?}");
            assert!((p[1] - (30.0 + DECAL_LIFT)).abs() < 1e-4, "height: {p:?}");
            assert!((p[2] - 20.0).abs() <= DECAL_SIZE, "depth: {p:?}");
        }
    }

    #[test]
    fn a_mark_is_lifted_off_its_own_surface() {
        // Coplanar with the wall it is on, a decal z-fights — which reads as
        // flicker rather than as a bug, so it is never reported as one.
        // Render y, not cube z: this reads a mark on a *floor*, whose height is
        // the axis it is lifted along.
        let mut pool = DecalPool::default();
        pool.mark([5.0, 5.0, 0.0], FACE_PZ);
        let mut heights = Vec::new();
        pool.quads(|p, _, _| heights.push(p[1]));
        assert_eq!(heights.len(), 6);
        for z in heights {
            assert!((z - DECAL_LIFT).abs() < 1e-6, "mark sat at {z}");
        }
    }

    #[test]
    fn a_mark_lies_flat_in_its_own_face() {
        // A mark on a wall must have no extent along that wall's normal, or it
        // is a box rather than a decal.
        let mut pool = DecalPool::default();
        pool.mark([5.0, 5.0, 3.0], FACE_NX);
        let mut xs = Vec::new();
        pool.quads(|p, _, _| xs.push(p[0]));
        let lo = xs.iter().copied().fold(f32::MAX, f32::min);
        let hi = xs.iter().copied().fold(f32::MIN, f32::max);
        assert!((hi - lo).abs() < 1e-6, "the mark has depth: {lo} to {hi}");
    }

    #[test]
    fn a_mark_is_the_size_it_says_it_is() {
        let mut pool = DecalPool::default();
        pool.mark([5.0, 5.0, 0.0], FACE_PZ);
        let mut xs = Vec::new();
        pool.quads(|p, _, _| xs.push(p[0]));
        let lo = xs.iter().copied().fold(f32::MAX, f32::min);
        let hi = xs.iter().copied().fold(f32::MIN, f32::max);
        assert!((hi - lo - DECAL_SIZE).abs() < 1e-6);
    }

    #[test]
    fn a_new_map_forgets_the_last_one() {
        let mut pool = DecalPool::default();
        pool.mark([0.0; 3], FACE_NX);
        pool.clear();
        assert!(pool.is_empty());
    }
}
